util/deephash/deephash.go

```diff
@@ -306,33 +292,40 @@ func (h *hasher) hashMapAcyclic(v reflect.Value) (acyclic bool) {
 		key := iterKey(iter, k)
 		val := iterVal(iter, e)
 		mh.startEntry()
-		if !h.print(key) {
-			return false
-		}
-		if !h.print(val) {
-			return false
-		}
+		h.print(key)
+		h.print(val)
 		mh.endEntry()
 	}
 	oldw.Write(mh.xbuf[:])
-	return true
 }
 
-func (h *hasher) hashMapFallback(v reflect.Value) (acyclic bool) {
-	acyclic = true
-	sm := newSortedMap(v)
-	w := h.bw
-	fmt.Fprintf(w, "map[%d]{\n", len(sm.Key))
-	for i, k := range sm.Key {
-		if !h.print(k) {
-			acyclic = false
-		}
-		w.WriteString(": ")
-		if !h.print(sm.Value[i]) {
-			acyclic = false
-		}
-		w.WriteString("\n")
+// visitStack is a stack of pointers visited.
+// Pointers are pushed onto the stack when visited, and popped when leaving.
+// The integer value is the depth at which the pointer was visited.
+// The length of this stack should be zero after every hashing operation.
+type visitStack map[pointer]int
+
+func (v visitStack) seen(p pointer) (int, bool) {
+	idx, ok := v[p]
+	return idx, ok
+}
+
+func (v *visitStack) push(p pointer) {
+	if *v == nil {
+		*v = make(map[pointer]int)
 	}
-	w.WriteString("}\n")
-	return acyclic
+	(*v)[p] = len(*v)
+}
+
+func (v visitStack) pop(p pointer) {
+	delete(v, p)
+}
+
+// pointer is a thin wrapper over unsafe.Pointer.
+// We only rely on comparability of pointers; we cannot rely on uintptr since
+// that would break if Go ever switched to a moving GC.
+type pointer struct{ p unsafe.Pointer }
+
+func pointerOf(v reflect.Value) pointer {
+	return pointer{unsafe.Pointer(v.Pointer())}
```

Comment:
Is it safe to do this? This is converting from a uintptr to an unsafe.Pointer, which is not on the list of allowed unsafe.Pointer operations. Does checkptr complain?

Reply:
> which is not on the list of allowed unsafe.Pointer operations.

On https://pkg.go.dev/unsafe#Pointer, I see:

> (5) Conversion of the result of reflect.Value.Pointer or reflect.Value.UnsafeAddr from uintptr to Pointer.

Doesn't that cover this case?

Reply:
Yes! Thanks. :) LGTM.